src/wasm/wasm-emscripten.cpp

```diff
@@ -702,45 +717,110 @@ void AsmConstWalker::noteNonLinear(Expression* curr) {
+      assert(value->op == AddInt32);
+      arg = value->right;
+    } else {
+      Fatal() << "Unexpected arg0 type (" << getExpressionName(arg)
+              << ") in call to to: " << target;
+    }
+  }
+  return arg->cast<Const>();
+}
+
+Index AsmConstWalker::resolveConstIndex(
+  Expression* arg, std::function<void(Expression*)> reportError) {
+  while (!arg->dynCast<Const>()) {
+    if (auto* get = arg->dynCast<LocalGet>()) {
+      // The argument may be a local.get, in which case, the last set in this
+      // basic block has the value.
+      auto* set = sets[get->index];
+      if (set) {
+        assert(set->index == get->index);
+        arg = set->value;
+      } else {
+        reportError(arg);
+        return 0;
+      }
+    } else if (arg->is<GlobalGet>()) {
+      // In the dynamic linking case, indices start at __table_base.
+      // We want the value relative to __table_base.
+      return 0;
+    } else {
+      reportError(arg);
+      return 0;
+    }
+  }
+  return Index(arg->cast<Const>()->value.geti32());
+}
+
 void AsmConstWalker::visitCall(Call* curr) {
   auto* import = wasm.getFunction(curr->target);
+  if (!import->imported()) {
+    return;
+  }
+
   // Find calls to emscripten_asm_const* functions whose first argument is
   // is always a string constant.
-  if (import->imported() && import->base.hasSubstring(EMSCRIPTEN_ASM_CONST)) {
+  if (import->base.hasSubstring(EMSCRIPTEN_ASM_CONST)) {
     auto baseSig = getSig(curr);
     auto sig = fixupNameWithSig(curr->target, baseSig);
-    auto* arg = curr->operands[0];
-    while (!arg->dynCast<Const>()) {
-      if (auto* get = arg->dynCast<LocalGet>()) {
-        // The argument may be a local.get, in which case, the last set in this
-        // basic block has the value.
-        auto* set = sets[get->index];
-        if (set) {
-          assert(set->index == get->index);
-          arg = set->value;
-        }
-      } else if (auto* value = arg->dynCast<Binary>()) {
-        // In the dynamic linking case the address of the string constant
-        // is the result of adding its offset to __memory_base.
-        // In this case are only looking for the offset with the data segment so
-        // the RHS of the addition is just what we want.
-        assert(value->op == AddInt32);
-        arg = value->right;
-      } else {
-        if (!value) {
-          Fatal() << "Unexpected arg0 type (" << getExpressionName(arg)
-                  << ") in call to to: " << import->base;
-        }
-      }
-    }
-
-    auto* value = arg->cast<Const>();
+    auto* value = resolveConstAddr(curr->operands[0], import->base);
     auto code = codeForConstAddr(wasm, segmentOffsets, value);
     sigsForCode[code].insert(sig);
 
     // Replace the first argument to the call with a Const index
     Builder builder(wasm);
     curr->operands[0] = builder.makeConst(idLiteralForCode(code));
+  } else if (import->base.startsWith(INVOKE_PREFIX)) {
```

Comment:
This code dealing with `INVOKE_PREFIX`es seems to be new. Can you add a comment explaining what is happening here and why?